spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RecurrentDecoder.scala

```diff
@@ -0,0 +1,217 @@
+     */
+    var i = 1
+    // init state
+    currentInput(hidDim) = if (initState != null) initState
+     else hidden
+
+    while (i <= times) {
+      // input at t(0) is user input
+      val inputTmp = if (i == 1) {
+        input
+      } else {
+        // input at t(i) is output at t(i-1)
+        cells(i - 2).output.toTable[Tensor[T]](inputDim)
+      }
+
+      currentInput(inputDim) = if (preTopology != null) {
+        newInput.narrow(2, i, 1).copy(inputTmp)
+        val sizes = 1 +: inputTmp.size()
+        inputTmp.resize(sizes)
+        val _input = preTopology.updateOutput(inputTmp).toTensor[T]
+        inputTmp.resize(sizes.takeRight(sizes.length - 1))
+        _input.select(1, 1)
+      } else {
+        outputCell.narrow(2, i, 1).copy(inputTmp)
+        inputTmp
+      }
+      cells(i - 1).updateOutput(currentInput)
+      currentInput(hidDim) = cells(i - 1).output.toTable(hidDim)
+      i += 1
+    }
+
+    if (preTopology != null) {
+      // For backward preTopology use
+      outputCell = preTopology.updateOutput(newInput).toTensor[T]
+    }
+
+    Recurrent.copy(cells.map(x => x.output.toTable[Tensor[T]](inputDim)), output)
+    output
+  }
+
+  override def accGradParameters(input: Tensor[T], gradOutput: Tensor[T]): Unit = {
+    throw new Exception("Should not enter RecurrentDecoder accGradParameters" +
+      "as it has override backward")
+  }
+
+  override def updateGradInput(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    throw new Exception("Should not enter RecurrentDecoder updateGradInput" +
+      "as it has override backward")
+    gradInput
+  }
+
+  override def backward(input: Tensor[T], gradOutput: Tensor[T]): Tensor[T] = {
+    val st = System.nanoTime
+    currentGradOutput(hidDim) = gradHidden
+    var i = times
+    while (i >= 1) {
+      currentGradOutput(inputDim) = if (i == times) gradOutput.select(timeDim, i)
+      else {
+        gradOutput.select(timeDim, i).clone()
+          .add(cells(i).gradInput.toTable[Tensor[T]](inputDim).clone())
+      }
+
+      _input(hidDim) = if (i > 1) cells(i - 2).output.toTable(hidDim)
+      else if (initState == null) hidden else initState
+      _input(inputDim) = Recurrent.selectCopy(outputCell, i, outputBuffer)
+
+      if (i == 1) {
+        cells(i - 1).regluarized(true)
+      } else {
+        cells(i - 1).regluarized(false)
+      }
+      cells(i - 1).backward(_input, currentGradOutput)
+      currentGradOutput(hidDim) = cells(i - 1).gradInput.toTable(hidDim)
+      i -= 1
+    }
+
+    gradInput = if (preTopology != null) {
+      /**
+       * if preTopology is Sequential, it has not created gradInput.
+       * Thus, it needs to create a new Tensor.
+       */
+      if (preTopology.gradInput == null) {
+        preTopology.gradInput = Tensor[T]()
+      }
+      preTopology.gradInput.toTensor[T]
+    } else {
+      gradInputCell
+    }
+    gradInputCell.resizeAs(outputCell)
```

Comment:
I don't think this is correct if preTopology == null: outputCell is B*T*H, and so gradInput is B*T*H, which is different from input (which is B*H)